kernel/kernel-api/src/main/java/io/delta/kernel/internal/DeltaErrors.java

```diff
@@ -42,4 +42,37 @@ public static RuntimeException nonExistentVersionException(
             latestVersion);
         return new RuntimeException(message);
     }
+
+    // TODO update to be user-facing exception with future exception framework
+    //  (see delta-io/delta#2231) & document in method docs as needed
+    //  (Table::getSnapshotAtTimestamp)
+    public static RuntimeException timestampEarlierThanTableFirstCommitException(
+            String tablePath, long providedTimestamp, long commitTimestamp) {
+        String message = String.format(
+            "%s: The provided timestamp %s is before the earliest version available. " +
+                "Please use a timestamp greater than or equal to %s",
+            tablePath,
+            providedTimestamp,
+            commitTimestamp);
+        // TODO format the timestamps?
```

Comment:
I would print both, the millis and a human readable one. See how delta-spark formats timestamps (if they are ever in millis format). I'd expect you can print them in the local time zone.

Reply:
I think this is a good thing to discuss when we revisit exceptions. I agree formatting them somehow makes sense but I'm not sure it's clear whether it should be in the local timezone or UTC and what we might expect the connectors to do with the exception. I'll add them in UTC for now.